PHC Binding - initial contribution

gnlpfjh opened the pull request:
Initial implementation of the PHC modules.
This binding allows you to integrate modules(at the Moment AM, EM and JRM) of PHC, without the PHC control (STM), in openHAB.

kubawolanin commented on addons/binding/org.openhab.binding.phc/ESH-INF/thing/channel-types.xml:
```diff
@@ -0,0 +1,85 @@
+        <description>Channel to an JRM module.</description>
+    </channel-type>
+ <!-- extra?? -->
+    <channel-type id="window-channel">
```
If it's extra, add `advanced="true"` parameter

kubawolanin commented on addons/binding/org.openhab.binding.phc/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,28 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .
+Import-Package: 
+ com.google.common.cache;version="10.0.1",
```
Please remove version constraints from the imported packages

kubawolanin commented on addons/binding/org.openhab.binding.phc/pom.xml:
```diff
@@ -0,0 +1,19 @@
+  <parent>
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.0.0-SNAPSHOT</version>
```
Please change to `2.1.0-SNAPSHOT`

kubawolanin commented on addons/binding/org.openhab.binding.phc/pom.xml:
```diff
@@ -0,0 +1,19 @@
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.phc</artifactId>
+  <version>2.0.0-SNAPSHOT</version>
```
You can remove `version` and `groupId` as it'll be inherited from `parent` node

kubawolanin commented:
Please also add [a proper about.html file](http://docs.openhab.org/developers/development/guidelines.html#b-osgi-bundles)

gnlpfjh commented:
I´ve corrected/updated these and some other things.
@kubawolanin

gnlpfjh commented on addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java:
```diff
@@ -0,0 +1,710 @@
+    RXTXPort serialPort = null;
+
+    private Map<String, Boolean> toggleMap = new HashMap<String, Boolean>();
+    private ConcurrentSkipListMap<Long, QueueObject> queue = new ConcurrentSkipListMap<Long, QueueObject>();
```
The method subMap (line 278) needs the `ConcurrentSkipListMap`. Maybe you have a better solution.

martinvw replied:
The most broad interface you could use then is `ConcurrentNavigableMap` [source](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentSkipListMap.html#subMap(K,%20boolean,%20K,%20boolean))

gnlpfjh commented on addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java:
```diff
@@ -0,0 +1,710 @@
+                sendPorBroadcast((byte) 0xFF);
+
+                byte[] b = { 0x01 };
+                for (int j = 0; j <= 0x1F; j++) { // TODO: move to PHCHandler?
```
I removed the TODO, it isn't sensible.

gnlpfjh commented on addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java:
```diff
@@ -0,0 +1,710 @@
+                    } else if (qo.getCounter() >= SEND_RETRY_COUNT
+                            && !qo.getModuleType().equals(PHCBindingConstants.CHANNELS_JRM)) {
+                        // Can´t process the acknowledgement of JRM yet.
+                        // TODO: Thing state to offline?
```
I think the serial communication isn't enough reliable. I wrote more detail in the TODO.

gnlpfjh commented on addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java:
```diff
@@ -0,0 +1,710 @@
+            while (portList.hasMoreElements()) {
+                CommPortIdentifier id = (CommPortIdentifier) portList.nextElement();
+                if (id.getPortType() == CommPortIdentifier.PORT_SERIAL) {
+                    if (id.getName().equals(port)) {
```
I get the name from the configuration. I saw it in another binding (RME Binding).

martinvw replied:
I think that example is not the most optimal, our codebase also contains the following which should be prefered (if it works ;-))

```java
CommPortIdentifier.getPortIdentifier(portName);
```

martinvw commented:
@gnlpfjh thanks for the update, If you are ready with it please let me know.

gnlpfjh commented:
Thank you, I've added both and it's working well. In addition I've added the configuration and sending for shutter duration.
@martinvw Now i'm ready.

martinvw commented:
Can you squash the last **7** (!!!!) commits and change commit message to include a `Signed-off-by` (to certify that you wrote it or otherwise have the right to pass it on as an open-source), see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work

For squashing / rebasing see:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

gnlpfjh commented:
@martinvw That doesn't look good. I used reword because the commit update fork is an merge commit too. I have the version bevore rebase on my PC so I can sqashing it(whith the "update fork" commit?).

martinvw commented:
It does not look good at all, merging is not really compatible with squash/merge and rebase strategy we use in this project, recovering from this is really though. The best thing to do, might be to a make patch for all your changes and apply that to a fresh master. For now it might be better to push previous state as it was.

gnlpfjh commented:
I would push the state before the rebase. Should I open a new PR and push all changes in one commit? I think that would be the easyest.

martinvw commented:
Yes and please refer to this one in case we forget :-)

gnlpfjh commented:
The new clean PR is #2255.
